src/lib.rs

```diff
@@ -424,30 +424,50 @@ impl<R: gimli::Reader> FullContext<R> {
 
 type Error = gimli::Error;
 
-fn str_attr<'abbrev, 'unit, R: gimli::Reader>(
+fn name_attr<'abbrev, 'unit, R: gimli::Reader>(
     entry: &gimli::DebuggingInformationEntry<'abbrev, 'unit, R, R::Offset>,
     dw_unit: &gimli::CompilationUnitHeader<R, R::Offset>,
     abbrevs: &gimli::Abbreviations,
     sections: &DebugSections<R>,
-    name: gimli::DwAt,
 ) -> Result<Option<R>, Error> {
-    Ok(match entry.attr(name)? {
-        Some(x) => x.string_value(&sections.debug_str),
-        None => {
-            match entry.attr_value(gimli::DW_AT_abstract_origin)? {
-                Some(gimli::AttributeValue::UnitRef(offset)) => {
-                    let mut tcursor = dw_unit.entries_at_offset(abbrevs, offset)?;
-                    match tcursor.next_dfs()? {
-                        // FIXME: evil dwarf can send us into an infinite loop here
-                        Some((_, entry)) => str_attr(entry, dw_unit, abbrevs, sections, name)?,
-                        None => None,
-                    }
-                }
-                None => None,
-                x => panic!("wat {:?}", x),
+    if let Some(attr) = entry.attr(gimli::DW_AT_linkage_name)? {
+        if let Some(val) = attr.string_value(&sections.debug_str) {
+            return Ok(Some(val));
+        }
+    }
+    if let Some(attr) = entry.attr(gimli::DW_AT_MIPS_linkage_name)? {
+        if let Some(val) = attr.string_value(&sections.debug_str) {
+            return Ok(Some(val));
+        }
+    }
+    if let Some(attr) = entry.attr(gimli::DW_AT_name)? {
+        if let Some(val) = attr.string_value(&sections.debug_str) {
+            return Ok(Some(val));
+        }
+    }
```

Comment:
I don't like the code duplication here. Perhaps something like `[gimli::DW_AT_linkage_name, gimli::DW_AT_MIPS_linkage_name, gimli::DW_AT_name].iter().filter_map(|x| entry.attr(x)).filter_map(|x| x.string_value(&sections.debug_str)).next()` helps?

Reply:
`DW_AT_linkage_name` and `DW_AT_MIPS_linkage_name` will be combined when we do something like [this commit](https://github.com/gimli-rs/addr2line/pull/59/commits/25471236e22d7a33a6307df915eee73ae4f003b7). Might be hard to completely avoid duplication for `DW_AT_name` though.